### PR TITLE
SparseLSTM

### LINKED ISSUE #1434: Sparse Vanilla LSTM
Hi @neubig,

I'm trying to implement a WaveRNN model, for a TTS-Project I'm working on (https://tiberiu44.github.io/TTS-Cube/). In order to make it work in realtime, I've already created a C++ backend, which is compatible with the model format stored by DyNet. This is to avoid overhead during the sampling procedure at runtime. I now need to go for the weight pruning technique described here: https://arxiv.org/pdf/1704.05119.pdf. For what I've read, there is no supporting function from this already in place.

Looking over the DyNet code, it seems pretty straight-forward to do this. I just need to create two mask vectors in the VanillaLSTM class for `p_x2i` and `p_h2i`, which I need to add to the `local_model`. Then I just use component-wise multiplication in the forward pass. Just to be sure, these parameters will be saved automatically in the model, right. I don't need to do anything else. 

I have two questions:
1) How do I modify the python bindings to add a new method for setting the pruning percentage;
2) Do you think this will be useful to others (should I create a branch and a pull request, once I'm finished), or do you consider it adds to much overhead to be worthwhile.

### PR DESCRIPTION
**Pull request by tiberiu44**
Hi,

I added a new `SparseLSTMBuilder`. It basically implements the weight pruning technique described in the WaveRNN paper (https://arxiv.org/pdf/1802.08435.pdf).

It adds a new method called `set_sparisty(float percent)`, which modifies the binary masks that are applied on the internal parameters. 

Closes #1434

### CODE REVIEW

**Comment by pmichel31415** [comment]
Looks like compilation is failing on travis, can you take a look?

**Comment by tiberiu44** [comment]
Yes, I started preparing a Linux box. Everything seems to work on OSX.

**Comment by tiberiu44** [comment]
Hi @pmichel31415,
It seems that the pull request now passes all checks. Sorry for the delay. It took me some time to get the right settings. On Debian it worked with the latest g++ and also with the oldest installable version (5.0).

File: dynet/lstm.h [hunk 1]
```diff
@@ -317,6 +318,157 @@ struct VanillaLSTMBuilder : public RNNBuilder {
...
+  /**
+   * \brief Set all dropout rates to 0
+   * \details This is equivalent to `set_dropout(0)` or `set_dropout(0,0,0)`
+   *
+   */
+   
+   void set_sparsity(float percent);
+   /**
+   * \brief Select the percent of weights that will be masked, based on their magnitude
+   */
```
Comment (pmichel31415) [inline dynet/lstm.h]: I think the docstrings for the two methods here (set_sparsity and disable_dropout) are interchanged.

File: dynet/lstm.h [hunk 2]
```diff
@@ -317,6 +318,157 @@ struct VanillaLSTMBuilder : public RNNBuilder {
...
+  /**
+   * \brief Set dropout masks at the beginning of a sequence for a specific batch size
+   * \details If this function is not called on batched input, the same mask will be applied across
+   * all batch elements. Use this to apply different masks to each batch element
+   *
+   * \param batch_size Batch size
+   */
+  void set_dropout_masks(unsigned batch_size = 1);
+  /**
+   * \brief Get parameters in VanillaLSTMBuilder
```
Comment (pmichel31415) [inline dynet/lstm.h]: VanillaLSTM -> SparseLSTM

File: python/_dynet.pyx [hunk 3]
```diff
@@ -5401,6 +5400,176 @@ cdef class VanillaLSTMBuilder(_RNNBuilder): # {{{
...
+
+    Args:
+        layers (int): Number of layers
+        input_dim (int): Dimension of the input
+        hidden_dim (int): Dimension of the recurrent units
+        model (dynet.ParameterCollection): ParameterCollection to hold the parameters
+        ln_lstm (bool): Whether to use layer normalization
+        forget_bias (float): value to use as forget gate bias(default 1.0)
+    """
+    cdef CSparseLSTMBuilder* thisvanillaptr
```
Comment (pmichel31415) [inline python/_dynet.pyx]: This variable should probably be renamed for readability

**Comment by tiberiu44** [comment]
I think I need some help. I don't understand why the gradient check failed. I just added two variables for mask vectors and did a `cmult` with the internal LSTM parameters. Initially, all the values are set to 1.0f. Also, the masks are constant parameters. 

When the masks are set to 1.0, the operations should be identical to the VanillaLSTM. I checked this manually when I was testing the code. Training the SparseLSTM with sparsity set to 0, generated identical output to that of the VanillaLSTM (in terms of loss and weights).

**Comment by pmichel31415** [comment]
OK so my educated guess is that check_grad doesn't handle const_parameters correctly and expects the gradient of the masks to be non-zero.

I don't think there is a straight-forward solution to this apart from fixing check-grad.cc which is a bit out of the scope of this PR so for now I suggest we remove the c++ test and add a disclaimer somewhere in the documentation that this is experimental/untested

**Comment by tiberiu44** [comment]
Hi @pmichel31415 ,
Is everything ok with this pull request? I think I covered all the requested changes.

**Comment by pmichel31415** [comment]
Oops, yes I'll merge, thanks

**Pull request merged**